File: android/app/src/main/kotlin/org/gaindrive/android/ui/browse/AlbumsViewModel.kt
package org.gaindrive.android.ui.browse

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.navigation.toRoute
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.gaindrive.android.data.CoverUrls
import org.gaindrive.android.data.LibraryRepository
import org.gaindrive.android.data.ServerFailure
import org.gaindrive.android.data.SettingsStore
import org.gaindrive.android.data.model.Album
import org.gaindrive.android.data.model.AlbumSort
import org.gaindrive.android.data.model.ArtistInfo
import org.gaindrive.android.data.model.ItemRef
import org.gaindrive.android.data.model.LibrarySection
import org.gaindrive.android.data.model.comparator
import org.gaindrive.android.net.runCatchingCancellable
import org.gaindrive.android.net.userMessage
import org.gaindrive.android.ui.AlbumUi
import org.gaindrive.android.ui.Load
import org.gaindrive.android.ui.Route
import javax.inject.Inject

/**
 * The artist header. Held separately from the album list because the server
 * may reach out to MusicBrainz and Wikipedia to build it, which can take
 * seconds or fail — neither of which may delay the albums.
 */
data class ArtistHeaderUi(
	val portraitUrl: String? = null,
	val info: ArtistInfo? = null,
)

@HiltViewModel
class AlbumsViewModel @Inject constructor(
	private val library: LibraryRepository,
	private val settings: SettingsStore,
	savedStateHandle: SavedStateHandle,
) : ViewModel() {

	private val route = savedStateHandle.toRoute<Route.Albums>()

	/**
	 * Every server's id for this artist. More than one when the row that was
	 * tapped merged artists of the same name.
	 */
	val artistRefs: List<ItemRef> = ItemRef.decodeAll(route.artistRefs)
		.ifEmpty { error("Bad artist refs: ${route.artistRefs}") }

	/**
	 * The first contributor in registry order. Its server answers for the
	 * portrait and biography — those are one server's opinion of the artist,
	 * and showing two of them stacked would be worse than picking one.
	 */
	private val primaryRef: ItemRef = artistRefs.first()

	val artistName: String = route.artistName

	/** Passed down to each album, and only ever true when browsing Uploads. */
	val fromUploads: Boolean = route.fromUploads

	/**
	 * True when this "artist" is a section of a categories root. It has no
	 * portrait and no biography — `is_category_folder()` on the server refuses
	 * the lookup — so the header draws neither, and neither is asked for.
	 */
	val isCategory: Boolean = route.fromCategories

	private val _albums = MutableStateFlow<Load<List<AlbumUi>>>(Load.Loading)

	/**
	 * Which section this listing was drilled in from, keying the sort
	 * preference: a discography wants a different order than a film category.
	 * Derived from the route flags rather than from any stored "current
	 * section", so the key always matches the listing on screen.
	 */
	private val section = when {
		route.fromUploads -> LibrarySection.UPLOADS
		route.fromCategories -> LibrarySection.CATEGORIES
		else -> LibrarySection.ARTISTS
	}

	val sort: StateFlow<AlbumSort> = settings.albumSort(section)
		.stateIn(viewModelScope, SharingStarted.Lazily, AlbumSort.DEFAULT)

	/**
	 * The listing, in the order chosen for this library slice.
	 *
	 * Sorted here rather than in [loadAlbums] so changing the order costs
	 * nothing on the wire: the whole list is already in hand, and re-reading it
	 * to reorder it would put a spinner over a decision that should be
	 * instant. It also fixes the order of a *merged* artist, whose albums
	 * arrive as one server's list concatenated with the next's.
	 */
	val state: StateFlow<Load<List<AlbumUi>>> =
		combine(_albums, sort) { load, order ->
			if (load is Load.Ready) {
				Load.Ready(
					load.value.sortedWith(compareBy<AlbumUi, Album>(order.comparator) { it.album })
				)
			} else {
				load
			}
		}.stateIn(viewModelScope, SharingStarted.Lazily, Load.Loading)

	fun setSort(order: AlbumSort) =
		viewModelScope.launch { settings.setAlbumSort(section, order) }

	private val _header = MutableStateFlow(ArtistHeaderUi())
	val header: StateFlow<ArtistHeaderUi> = _header.asStateFlow()

	private val _failures = MutableStateFlow<List<ServerFailure>>(emptyList())
	val failures: StateFlow<List<ServerFailure>> = _failures.asStateFlow()

	/** True only for a user-initiated pull, which drives the pull indicator. */
	private val _isRefreshing = MutableStateFlow(false)
	val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

	init {
		load()
	}

	/** Initial load and retry: there is nothing worth keeping on screen. */
	fun load() {
		_albums.value = Load.Loading
		loadAlbums()
		loadHeader()
	}

	/** Pull to refresh: keep the albums visible while they are re-read. */
	fun refresh() {
		_isRefreshing.value = true
		loadAlbums()
		loadHeader()
	}

	fun dismissFailures() {
		_failures.value = emptyList()
	}

	private fun loadAlbums() = viewModelScope.launch {
		runCatchingCancellable {
			// Resolved once for the whole list, not per row.
			val covers: CoverUrls = library.coverUrls()
			// Badges here follow the artist, not the browse scope: this row
			// list is a union across the servers the merged artist came from,
			// and a single-server artist needs no badge on every row.
			val names =
				if (artistRefs.size < 2) emptyMap()
				else library.enabledServers().associate { it.id to it.name }

			library.albumsOfArtist(artistRefs).map { albums ->
				albums.map { album ->
					AlbumUi(
						album = album,
						coverUrl = covers.url(album.coverArt, COVER_PX),
						badges = album.sources.mapNotNull { names[it] },
					)
				}
			}
		}.fold(
			onSuccess = { merged ->
				if (merged.items.isEmpty() && merged.isPartial) {
					_albums.value = Load.Failed(merged.failures.first().message)
				} else {
					_albums.value = Load.Ready(merged.items)
				}
				_failures.value = merged.failures
			},
			onFailure = {
				_albums.value = Load.Failed(it.userMessage())
				_failures.value = emptyList()
			},
		)
		_isRefreshing.value = false
	}

	/**
	 * Fills in the header as it arrives: the portrait URL first, then the
	 * biography, which may involve a lookup.
	 *
	 * Note that the portrait URL arriving is not the portrait arriving. Both
	 * halves of this header come from a server-side lookup now — `getCoverArt`
	 * answers 404 for an artist it has not resolved yet, and only the request
	 * itself puts them at the front of the queue. Re-running this function
	 * cannot recover from that, because the URL it computes is identical and
	 * [ArtistHeaderUi] is a data class, so the state never changes and nothing
	 * recomposes. The retry lives in `ArtistAvatar`, which is the only thing
	 * that can see whether the image actually loaded.
	 */
	private fun loadHeader() = viewModelScope.launch {
		// Nothing to fill in for a section, and nothing to wait for: the
		// portrait would 404 for as long as ArtistAvatar kept retrying it, and
		// getArtistInfo2 answers empty by design.
		if (isCategory) return@launch

		runCatchingCancellable {
			val covers = library.coverUrls()
			covers.url(primaryRef, PORTRAIT_PX)
		}.getOrNull()?.let { url ->
			_header.update { it.copy(portraitUrl = url) }
		}

		// A missing biography is entirely normal and never worth an error.
		runCatchingCancellable { library.artistInfo(primaryRef) }
			.getOrNull()
			?.let { info -> _header.update { it.copy(info = info) } }
	}

	private companion object {
		/** Matches the 48dp thumbnail at roughly 3x density. */
		const val COVER_PX = 144

		/** Matches the 96dp avatar at roughly 3x density. */
		const val PORTRAIT_PX = 288
	}
}

File: ios/GainDrive/UI/Browse/AlbumsViewModel.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// One artist's albums, across every server that has them.
///
/// Owned by the screen rather than by `RootView`, because its state should die
/// with the screen: pushing a different artist is a different subject, not a
/// refresh of this one.
@MainActor
@Observable
final class AlbumsViewModel {
	private(set) var state: Load<[AlbumUi]> = .loading
	private(set) var failures: [ServerFailure] = []
	/// Held **separately from the list**, and never awaited before it.
	/// Answering `getArtistInfo2` may send the server out to MusicBrainz and
	/// Wikipedia, so a biography that is slow or missing must cost the
	/// biography and nothing else.
	private(set) var info: ArtistInfo?
	private(set) var portrait: CoverSource?
	private(set) var sort: AlbumSort

	@ObservationIgnored private let library: LibraryRepository
	@ObservationIgnored private let selection: ServerSelection
	@ObservationIgnored private let settings: SettingsStore
	@ObservationIgnored private let refs: [ItemRef]
	/// A section of a **categories** root: Film and Series are not performers,
	/// so there is no portrait to fetch and no biography to wait for.
	@ObservationIgnored private let fromCategories: Bool
	/// Which section's preference the sort belongs to, derived from the route
	/// flags the listing was drilled in from — this used to read a stored
	/// "current chip", which could disagree with the listing on screen.
	@ObservationIgnored private let section: LibrarySection
	@ObservationIgnored private var loaded = false
	@ObservationIgnored private var task: Task<Void, Never>?

	init(
		library: LibraryRepository, selection: ServerSelection, settings: SettingsStore,
		refs: [ItemRef], fromCategories: Bool, fromUploads: Bool = false
	) {
		self.library = library
		self.selection = selection
		self.settings = settings
		self.refs = refs
		self.fromCategories = fromCategories
		let section: LibrarySection =
			fromUploads ? .uploads : fromCategories ? .categories : .artists
		self.section = section
		self.sort = settings.albumSort(for: section)
	}

	/// **Re-sorted in the client, never asked of the server.** The whole list
	/// is already in hand, so the order changes with no request and no spinner.
	///
	/// It is also what fixes a merged artist's order, which was never in year
	/// order at all: `Merge.albums` keeps arrival order, so the union was one
	/// server's list concatenated with the next's.
	func setSort(_ next: AlbumSort) {
		guard next != sort else { return }
		sort = next
		settings.setAlbumSort(next, for: section)
		guard case .ready(let items) = state else { return }
		state = .ready(items.sorted { sort.precedes($0.album, $1.album) })
	}

	func appear() {
		guard !loaded else { return }
		loaded = true
		start(clearFirst: true)
	}

	func retry() {
		start(clearFirst: true)
	}

	func refresh() async {
		start(clearFirst: false)
		await task?.value
	}

	private func start(clearFirst: Bool) {
		if clearFirst { state = .loading }
		let badges = selection.badgeNames
		task?.cancel()
		task = Task { [library, refs] in
			let covers = await library.coverUrls()
			let merged = await library.albumsOfArtist(refs)
			guard !Task.isCancelled else { return }
			self.state = merged.map { albums in
				albums.map {
					AlbumUi(
						album: $0,
						cover: covers.source($0.coverArt, size: CoverSize.thumb),
						badges: $0.sources.compactMap { badges[$0] })
				}
				.sorted { self.sort.precedes($0.album, $1.album) }
			}.load
			self.failures = merged.failures

			// Started only once the list is on screen, and in its own task, so
			// it cannot delay anything.
			//
			// **Not at all for a categories section.** The server refuses the
			// lookup for one (`is_category_folder()`), so the avatar would sit
			// as a placeholder for ever while the fetch retried a 404 — and
			// `AlbumsView` draws no header when both of these stay nil.
			guard !self.fromCategories, let primary = refs.first else { return }
			self.portrait = covers.source(primary, size: CoverSize.portrait)
			self.info = await library.artistInfo(primary)
		}
	}

	func dismissFailures() {
		failures = []
	}
}
